File: Difficulty: Hard/Max Circular Subarray Sum/max-circular-subarray-sum.cpp
//{ Driver Code Starts
#include <bits/stdc++.h>

using namespace std;


// } Driver Code Ends
class Solution {
  public:
    // arr: input array
    // Function to find maximum circular subarray sum.
    int circularSubarraySum(vector<int> &arr) {
        // your code here
        int s=0;
        int sum=0;
        int res=0;
        for(auto it:arr)
        {
            sum+=it;
            s+=it;
            if(s<0)
            s=0;
            
            res=max(res,s);
        }
        s=0;
        int x=INT_MAX;
        for(auto it:arr)
        {
            s+=it;
            if(s>0)
            s=0;
            
            x=min(x,s);
        }
        res=max(res,sum-x);
        return res;
    }
};

//{ Driver Code Starts.
int main() {
    int t;
    cin >> t;
    cin.ignore();
    while (t--) {
        vector<int> arr;
        string input;

        // Read first array
        getline(cin, input);
        stringstream ss(input);
        int number;
        while (ss >> number) {
            arr.push_back(number);
        }

        Solution ob;
        int res = ob.circularSubarraySum(arr);

        cout << res << endl;
    }
    return 0;
}

// } Driver Code Ends

File: Difficulty: Hard/Max Circular Subarray Sum/max-circular-subarray-sum.java
class Solution {
    public int maxCircularSum(int a[]) {
        // code here
        int min = 0, max = Integer.MIN_VALUE, total = 0, minSum = 0, maxSum = 0;
        for(int ele : a) {
            total += ele;
            minSum += ele;
            maxSum += ele;

            min = Math.min(min, minSum);
            if(minSum > 0) minSum = 0;
            max = Math.max(max, maxSum);
            if(maxSum < 0) maxSum = 0;
        }
        
        if(max < 0) return max;
        return Math.max(max, (total - min));
    }
}
